File: catty-web/packages/terminal-emulation/src/terminal/TerminalTheme.ts
/**
 * Terminal Theme System
 * 
 * Provides a structured approach to color management using CSS custom properties.
 * Supports standard 16 ANSI colors plus terminal UI colors.
 */

import { DomStyleManager } from "./DomStyleManager";

/**
 * Terminal color palette containing all standard ANSI colors and UI colors
 */
export interface TerminalColorPalette {
  // Standard 16 ANSI colors
  black: string;
  red: string;
  green: string;
  yellow: string;
  blue: string;
  magenta: string;
  cyan: string;
  white: string;
  brightBlack: string;
  brightRed: string;
  brightGreen: string;
  brightYellow: string;
  brightBlue: string;
  brightMagenta: string;
  brightCyan: string;
  brightWhite: string;
  
  // Terminal UI colors
  foreground: string;
  background: string;
  cursor: string;
  selection: string;
}

/**
 * Complete terminal theme definition
 */
export interface TerminalTheme {
  name: string;
  type: "dark" | "light";
  colors: TerminalColorPalette;
}

/**
 * Default dark theme using standard terminal colors
 */
export const DEFAULT_DARK_THEME: TerminalTheme = {
  name: "Default Dark",
  type: "dark",
  colors: {
    // Standard ANSI colors
    black: "#000000",
    red: "#AA0000",
    green: "#00AA00",
    yellow: "#AA5500",
    blue: "#0000AA",
    magenta: "#AA00AA",
    cyan: "#00AAAA",
    white: "#AAAAAA",
    
    // Bright ANSI colors
    brightBlack: "#555555",
    brightRed: "#FF5555",
    brightGreen: "#55FF55",
    brightYellow: "#FFFF55",
    brightBlue: "#5555FF",
    brightMagenta: "#FF55FF",
    brightCyan: "#55FFFF",
    brightWhite: "#FFFFFF",
    
    // Terminal UI colors
    foreground: "#AAAAAA",
    background: "#000000",
    cursor: "#AAAAAA",
    selection: "#444444",
  },
};

/**
 * Maps ANSI color names to CSS variable names
 */
const COLOR_VARIABLE_MAP: Record<keyof TerminalColorPalette, string> = {
  black: "--terminal-color-black",
  red: "--terminal-color-red",
  green: "--terminal-color-green",
  yellow: "--terminal-color-yellow",
  blue: "--terminal-color-blue",
  magenta: "--terminal-color-magenta",
  cyan: "--terminal-color-cyan",
  white: "--terminal-color-white",
  brightBlack: "--terminal-color-bright-black",
  brightRed: "--terminal-color-bright-red",
  brightGreen: "--terminal-color-bright-green",
  brightYellow: "--terminal-color-bright-yellow",
  brightBlue: "--terminal-color-bright-blue",
  brightMagenta: "--terminal-color-bright-magenta",
  brightCyan: "--terminal-color-bright-cyan",
  brightWhite: "--terminal-color-bright-white",
  foreground: "--terminal-foreground",
  background: "--terminal-background",
  cursor: "--terminal-cursor",
  selection: "--terminal-selection",
};

/**
 * Theme manager for handling terminal color themes and CSS variable generation
 */
export class ThemeManager {
  private currentTheme: TerminalTheme;

  constructor(initialTheme: TerminalTheme = DEFAULT_DARK_THEME) {
    this.currentTheme = initialTheme;
  }

  /**
   * Get the current active theme
   */
  public getCurrentTheme(): TerminalTheme {
    return this.currentTheme;
  }

  /**
   * Apply a new theme and generate CSS variables
   */
  public applyTheme(theme: TerminalTheme): void {
    this.currentTheme = theme;
    this.injectCssVariables();
  }

  /**
   * Generate CSS custom properties string for the current theme
   */
  public generateCssVariables(theme: TerminalTheme = this.currentTheme): string {
    const cssRules: string[] = [];
    
    for (const [colorName, colorValue] of Object.entries(theme.colors)) {
      const cssVariable = COLOR_VARIABLE_MAP[colorName as keyof TerminalColorPalette];
      if (cssVariable) {
        cssRules.push(`  ${cssVariable}: ${colorValue};`);
      }
    }
    
    return `:root {\n${cssRules.join('\n')}\n}`;
  }

  /**
   * Resolve a color by ANSI color code to CSS variable or direct value
   * @param colorCode ANSI color code (0-15 for standard colors)
   * @param isBright Whether this is a bright color variant
   * @returns CSS variable reference or direct color value
   */
  public resolveColor(colorCode: number, isBright: boolean = false): string {
    // Map ANSI color codes to color names
    const colorNames = [
      "black", "red", "green", "yellow", 
      "blue", "magenta", "cyan", "white"
    ];
    
    if (colorCode >= 0 && colorCode <= 7) {
      const colorName = colorNames[colorCode];
      const fullColorName = isBright ? `bright${colorName.charAt(0).toUpperCase()}${colorName.slice(1)}` : colorName;
      const cssVariable = COLOR_VARIABLE_MAP[fullColorName as keyof TerminalColorPalette];
      return `var(${cssVariable})`;
    }
    
    // For colors outside standard range, return a fallback
    return "var(--terminal-foreground)";
  }

  /**
   * Inject CSS variables into the DOM
   */
  private injectCssVariables(): void {
    // Check if we're in a browser environment
    if (typeof document === 'undefined') {
      return;
    }

    const cssText = this.generateCssVariables();
    DomStyleManager.getInstance().injectCss('terminal-theme-variables', cssText);
  }
}

File: catty-ksa/caTTY.Display/Rendering/TerminalTheme.cs
using Brutal.Numerics;
using caTTY.Core.Types;

namespace caTTY.Display.Rendering;

/// <summary>
/// Terminal color palette containing all standard ANSI colors and UI colors.
/// Based on TypeScript TerminalTheme implementation.
/// </summary>
public readonly struct TerminalColorPalette
{
    // Standard 16 ANSI colors
    public float4 Black { get; }
    public float4 Red { get; }
    public float4 Green { get; }
    public float4 Yellow { get; }
    public float4 Blue { get; }
    public float4 Magenta { get; }
    public float4 Cyan { get; }
    public float4 White { get; }
    public float4 BrightBlack { get; }
    public float4 BrightRed { get; }
    public float4 BrightGreen { get; }
    public float4 BrightYellow { get; }
    public float4 BrightBlue { get; }
    public float4 BrightMagenta { get; }
    public float4 BrightCyan { get; }
    public float4 BrightWhite { get; }
    
    // Terminal UI colors
    public float4 Foreground { get; }
    public float4 Background { get; }
    public float4 Cursor { get; }
    public float4 Selection { get; }

    public TerminalColorPalette(
        float4 black, float4 red, float4 green, float4 yellow,
        float4 blue, float4 magenta, float4 cyan, float4 white,
        float4 brightBlack, float4 brightRed, float4 brightGreen, float4 brightYellow,
        float4 brightBlue, float4 brightMagenta, float4 brightCyan, float4 brightWhite,
        float4 foreground, float4 background, float4 cursor, float4 selection)
    {
        Black = black;
        Red = red;
        Green = green;
        Yellow = yellow;
        Blue = blue;
        Magenta = magenta;
        Cyan = cyan;
        White = white;
        BrightBlack = brightBlack;
        BrightRed = brightRed;
        BrightGreen = brightGreen;
        BrightYellow = brightYellow;
        BrightBlue = brightBlue;
        BrightMagenta = brightMagenta;
        BrightCyan = brightCyan;
        BrightWhite = brightWhite;
        Foreground = foreground;
        Background = background;
        Cursor = cursor;
        Selection = selection;
    }
}

/// <summary>
/// Cursor configuration for terminal themes.
/// </summary>
public readonly struct CursorConfig
{
    /// <summary>
    /// Default cursor style for the theme.
    /// </summary>
    public CursorStyle DefaultStyle { get; }

    /// <summary>
    /// Whether cursor should blink by default.
    /// </summary>
    public bool DefaultBlink { get; }

    /// <summary>
    /// Cursor blink interval in milliseconds.
    /// </summary>
    public int BlinkIntervalMs { get; }

    public CursorConfig(CursorStyle defaultStyle, bool defaultBlink, int blinkIntervalMs = 500)
    {
        DefaultStyle = defaultStyle;
        DefaultBlink = defaultBlink;
        BlinkIntervalMs = blinkIntervalMs;
    }
}

/// <summary>
/// Complete terminal theme definition.
/// </summary>
public readonly struct TerminalTheme
{
    public string Name { get; }
    public ThemeType Type { get; }
    public TerminalColorPalette Colors { get; }
    public CursorConfig Cursor { get; }

    public TerminalTheme(string name, ThemeType type, TerminalColorPalette colors, CursorConfig cursor)
    {
        Name = name;
        Type = type;
        Colors = colors;
        Cursor = cursor;
    }
}

/// <summary>
/// Theme type enumeration.
/// </summary>
public enum ThemeType
{
    Dark,
    Light
}

/// <summary>
/// Theme manager for handling terminal color themes.
/// Provides default themes and color resolution.
/// </summary>
public static class ThemeManager
{
    /// <summary>
    /// Default dark theme using standard terminal colors.
    /// Matches the TypeScript DEFAULT_DARK_THEME.
    /// </summary>
    public static readonly TerminalTheme DefaultDarkTheme = new(
        "Default Dark",
        ThemeType.Dark,
        new TerminalColorPalette(
            // Standard ANSI colors
            black: new float4(0.0f, 0.0f, 0.0f, 1.0f),
            red: new float4(0.667f, 0.0f, 0.0f, 1.0f),
            green: new float4(0.0f, 0.667f, 0.0f, 1.0f),
            yellow: new float4(0.667f, 0.333f, 0.0f, 1.0f),
            blue: new float4(0.0f, 0.0f, 0.667f, 1.0f),
            magenta: new float4(0.667f, 0.0f, 0.667f, 1.0f),
            cyan: new float4(0.0f, 0.667f, 0.667f, 1.0f),
            white: new float4(0.667f, 0.667f, 0.667f, 1.0f),
            
            // Bright ANSI colors
            brightBlack: new float4(0.333f, 0.333f, 0.333f, 1.0f),
            brightRed: new float4(1.0f, 0.333f, 0.333f, 1.0f),
            brightGreen: new float4(0.333f, 1.0f, 0.333f, 1.0f),
            brightYellow: new float4(1.0f, 1.0f, 0.333f, 1.0f),
            brightBlue: new float4(0.333f, 0.333f, 1.0f, 1.0f),
            brightMagenta: new float4(1.0f, 0.333f, 1.0f, 1.0f),
            brightCyan: new float4(0.333f, 1.0f, 1.0f, 1.0f),
            brightWhite: new float4(1.0f, 1.0f, 1.0f, 1.0f),
            
            // Terminal UI colors
            foreground: new float4(0.667f, 0.667f, 0.667f, 1.0f),
            background: new float4(0.0f, 0.0f, 0.0f, 1.0f),
            cursor: new float4(0.667f, 0.667f, 0.667f, 1.0f),
            selection: new float4(0.267f, 0.267f, 0.267f, 1.0f)
        ),
        new CursorConfig(CursorStyle.BlinkingBlock, true, 500)
    );

    /// <summary>
    /// Default light theme for terminals.
    /// </summary>
    public static readonly TerminalTheme DefaultLightTheme = new(
        "Default Light",
        ThemeType.Light,
        new TerminalColorPalette(
            // Standard ANSI colors (darker for light background)
            black: new float4(0.0f, 0.0f, 0.0f, 1.0f),
            red: new float4(0.8f, 0.0f, 0.0f, 1.0f),
            green: new float4(0.0f, 0.6f, 0.0f, 1.0f),
            yellow: new float4(0.8f, 0.6f, 0.0f, 1.0f),
            blue: new float4(0.0f, 0.0f, 0.8f, 1.0f),
            magenta: new float4(0.8f, 0.0f, 0.8f, 1.0f),
            cyan: new float4(0.0f, 0.6f, 0.6f, 1.0f),
            white: new float4(0.8f, 0.8f, 0.8f, 1.0f),
            
            // Bright ANSI colors
            brightBlack: new float4(0.4f, 0.4f, 0.4f, 1.0f),
            brightRed: new float4(1.0f, 0.2f, 0.2f, 1.0f),
            brightGreen: new float4(0.2f, 0.8f, 0.2f, 1.0f),
            brightYellow: new float4(1.0f, 0.8f, 0.2f, 1.0f),
            brightBlue: new float4(0.2f, 0.2f, 1.0f, 1.0f),
            brightMagenta: new float4(1.0f, 0.2f, 1.0f, 1.0f),
            brightCyan: new float4(0.2f, 0.8f, 0.8f, 1.0f),
            brightWhite: new float4(1.0f, 1.0f, 1.0f, 1.0f),
            
            // Terminal UI colors (inverted for light theme)
            foreground: new float4(0.0f, 0.0f, 0.0f, 1.0f),
            background: new float4(1.0f, 1.0f, 1.0f, 1.0f),
            cursor: new float4(0.0f, 0.0f, 0.0f, 1.0f),
            selection: new float4(0.8f, 0.8f, 0.8f, 1.0f)
        ),
        new CursorConfig(CursorStyle.BlinkingBlock, true, 500)
    );

    /// <summary>
    /// Current active theme. Defaults to dark theme.
    /// </summary>
    public static TerminalTheme CurrentTheme { get; private set; } = DefaultDarkTheme;

    /// <summary>
    /// Apply a new theme.
    /// </summary>
    /// <param name="theme">The theme to apply</param>
    public static void ApplyTheme(TerminalTheme theme)
    {
        CurrentTheme = theme;
    }

    /// <summary>
    /// Resolve a color by ANSI color code using the current theme.
    /// </summary>
    /// <param name="colorCode">ANSI color code (0-15 for standard colors)</param>
    /// <returns>float4 color value</returns>
    public static float4 ResolveThemeColor(int colorCode)
    {
        var palette = CurrentTheme.Colors;
        
        return colorCode switch
        {
            0 => palette.Black,
            1 => palette.Red,
            2 => palette.Green,
            3 => palette.Yellow,
            4 => palette.Blue,
            5 => palette.Magenta,
            6 => palette.Cyan,
            7 => palette.White,
            8 => palette.BrightBlack,
            9 => palette.BrightRed,
            10 => palette.BrightGreen,
            11 => palette.BrightYellow,
            12 => palette.BrightBlue,
            13 => palette.BrightMagenta,
            14 => palette.BrightCyan,
            15 => palette.BrightWhite,
            _ => palette.Foreground
        };
    }

    /// <summary>
    /// Get the default foreground color from the current theme.
    /// </summary>
    public static float4 GetDefaultForeground() => CurrentTheme.Colors.Foreground;

    /// <summary>
    /// Get the default background color from the current theme.
    /// </summary>
    public static float4 GetDefaultBackground() => CurrentTheme.Colors.Background;

    /// <summary>
    /// Get the cursor color from the current theme.
    /// </summary>
    public static float4 GetCursorColor() => CurrentTheme.Colors.Cursor;

    /// <summary>
    /// Get the selection color from the current theme.
    /// </summary>
    public static float4 GetSelectionColor() => CurrentTheme.Colors.Selection;

    /// <summary>
    /// Get the default cursor style from the current theme.
    /// </summary>
    public static CursorStyle GetDefaultCursorStyle() => CurrentTheme.Cursor.DefaultStyle;

    /// <summary>
    /// Get the default cursor blink setting from the current theme.
    /// </summary>
    public static bool GetDefaultCursorBlink() => CurrentTheme.Cursor.DefaultBlink;

    /// <summary>
    /// Get the cursor blink interval from the current theme.
    /// </summary>
    public static int GetCursorBlinkInterval() => CurrentTheme.Cursor.BlinkIntervalMs;
}